R/methods-epi_df.R

```diff
@@ -383,3 +383,36 @@ arrange_canonical.epi_df <- function(x, ...) {
     dplyr::relocate(dplyr::all_of(keys), .before = 1) %>%
     dplyr::arrange(dplyr::across(dplyr::all_of(keys)))
 }
+
+#' Aggregate an `epi_df` object
+#'
+#' Aggregates an `epi_df` object by the specified group columns, summing the
+#' `value` column, and returning an `epi_df`. If aggregating over `geo_value`,
+#' the resulting `epi_df` will have `geo_value` set to `"total"`.
+#'
+#' @param .x an `epi_df`
+#' @param value_col character name of the column to aggregate
+#' @param group_cols character vector of column names to group by
+#' @return an `epi_df` object
+#'
+#' @export
+aggregate_epi_df <- function(.x, value_col = "value", group_cols = "time_value") {
+  assert_class(.x, "epi_df")
+  assert_character(value_col, len = 1)
+  assert_character(group_cols)
+  checkmate::assert_subset(value_col, names(.x))
+  checkmate::assert_subset(group_cols, names(.x))
+
+  .x %>%
+    group_by(across(all_of(group_cols))) %>%
+    dplyr::summarize(!!(value_col) := sum(!!sym(value_col))) %>%
+    ungroup() %>%
+    {
+      if (!"geo_value" %in% group_cols) {
+        mutate(., geo_value = "total") %>% relocate(geo_value, .before = 1)
+      } else {
+        .
+      }
+    } %>%
+    as_epi_df(as_of = attr(.x, "metadata")$as_of)
```

Comment:
issue: this probably aborts if `"time_value"` isn't in the `group_col`s; we should either
- require/auto-add it to the group cols or
- allow the result to be a tibble in this case.

issue: `other_keys` won't be set properly.
- If `group_cols` are a subset of `"geo_value"`, `"time_value"`, and `.x`'s `other_keys`, then the result's `other_keys` should be the non-geo-time grouping cols.
- If `group_cols` include other things, then I'm not sure what we'd do.  Easiest is to just not let them.  I'm trying to think when they might want to:
  - If user is trying to group by season/wave/quarter/wday: this seems plausible.  We might only want an `epi_df` as the result when aggregating by season, though, and then (something of an accepted time type standing in for) the season would become the time_value rather than the season becoming one of the `other_keys`.

Reply:
musing: it was a bit complicated to think about whether we get the right `geo_type`.  I think we do & this is how, but it seems convoluted.  Perhaps worth a source comment.
- if grouping by at least `geo_value` and `time_value`, then we should have the same `geo_type` as `.x`; we feed an `epi_df` with the right `geo_type` into `as_epi_df` and it trusts our metadata.
- if not grouping by `geo_value`, then the `geo_type` should be `"custom"`; we get this by `summarize` decaying to a tibble &`as_epi_df` setting the `geo_type` as "custom".

Reply:
issue 1: fixed. 
issue 2: i'm going to assume group_cols are a subset of key_colnames(.x) for now and make sure other_keys are set properly. the main goal of this function is to have a built-in workaround for people that wanted to use the old epi_slide as a way to aggregate geos while sliding (with this they can aggregate by non-geo_value and then slide). don't want to lose focus by getting into up/down scaling time resolution and up/down scaling geo resolution here.
musing: good point, will make sure we inherit the geo_type if geo_value is in group_cols

Reply:
side-note, we should figure out if we want to maintain this geo_type metadata inside an epi_df versus just writing some helper functions that can up/down scale the geo resolution and recognize geo_type, etc.